toolchain/diagnostics/format_providers.h

```diff
@@ -0,0 +1,135 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_DIAGNOSTICS_FORMAT_PROVIDERS_H_
+#define CARBON_TOOLCHAIN_DIAGNOSTICS_FORMAT_PROVIDERS_H_
+
+#include "common/check.h"
+#include "common/ostream.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/FormatVariadic.h"
+
+namespace Carbon {
+
+// Selects a formatv string based on the value.
+//
+// When used, the true and false outputs are separated by a `|`.
+//
+// For example, `{0:true|false}` would yield standard bool formatting.
+//
+// If needed, the _full_ style string can be wrapped with `'` in order to
+// preserve prefix or suffix whitespace (which is stripped by formatv). For
+// example, `{0:' true | false '}` retains whitespace which would be dropped
+// before `true` and after `false`.
```

Comment:
Seems like a reasonable workaround. But the need to do this makes me sad.

Reply:
Looks like std::format passes through the string with leading/trailing whitespace to the formatter.

So I'm happy to make the LLVM change to avoid trimming this so it works as desired. I'll follow-up here with the upstream PR/commit soon.

Reply:
We can update the implementation once LLVM has support; I'd rather not block this PR on that, to minimize merge conflicts.

Reply:
Posted upstream change here: https://github.com/llvm/llvm-project/pull/112625